### ISSUE TITLE
Sinon mock expectations do not remember multiple withArgs() or withExactArgs() calls.

### ISSUE THREAD

**Issue by TonyApuzzo** [sinonjs/sinon#992]
SInon 1.17.3 doesn't seem to work with mocks as expected.  I would think that I can treat mocks like stubs, but the with*Args() methods seem to only remember the last invocation, it doesn't create an internal map of argument to return value like you get with stubs.

``` javascript
describe('Sinon Verification', function () {
  it('Should work with mocks', function () {
    var u = { inc : function(x){} };
    var m = sinon.mock(u);
    var e = m.expects("inc");
    e.withExactArgs(0).returns(1);
    e.withExactArgs(1).returns(2);
    e.withExactArgs(2).returns(3);

    u.inc(2).must.eql(3); // OK
    u.inc(1).must.eql(2); // Fails expectation
    u.inc(0).must.eql(1); // Fails expectation
  });
});
```

**Comment by roderickhsiao** [comment]
having the same issue

**Comment by yaronyg** [comment]
Sounds like #984

**Comment by hrimhari** [comment]
@yaronyg not exactly. There were discussions about multiple withArgs() being chainable or not like #984 is trying to do. But the problem reported here is that mocks just can't have multiple withArgs(), regardless of their declaration being in the same chain or not.

**Comment by mroderick** [comment]
It is correct that mocks only keep the last value of arguments to `withArgs`.

This has been the case since 84dbb22ae72a94380b846ee58a7d2b0238c5ea0d, which was committed on Apr 18, 2010.

```shell
git log -S "this.expectedArguments = slice.call(arguments);"
```

The same is true for `withExactArgs`, which calls `withArgs`.

I have created https://github.com/sinonjs/sinon/issues/1432 to document this issue

**Comment by hrimhari** [comment]
@mroderick how would you test a loop over different objects if your mock can only accept the same parameters?

**Comment by mroderick** [comment]
> @mroderick how would you test a loop over different objects if your mock can only accept the same parameters?

With `sinon.mock`, you would have to create a new one for each member in your collection. Using `sinon.stub`, you can inspect each call.

**Comment by mroderick** [comment]
Issue #1193 also has a good explanation of this deficiency.

**Comment by mroderick** [comment]
Fixed by #1432
